src/ips.rs

```diff
@@ -0,0 +1,526 @@
+                peers_to_add_count = self.config.change_no_more;
+            }
+
+            // 4 - Choose peers to delete from peerlist (based on ranking)
+            if peers_to_delete_count > 0 {
+                // Sort peers by rating (highest first)
+                curr_peer_ratings.sort_by(|a, b| b.2.partial_cmp(&a.2).unwrap());
+
+                // Remove peers with lowest rating
+                for _ in 0..peers_to_delete_count {
+                    let peer_to_delete = curr_peer_ratings.pop();
+                    if let Some(peer_to_delete) = peer_to_delete {
+                        peer_list_entry.list.retain(|x| x != &peer_to_delete.0);
+                    }
+                }
+            }
+
+            // 5 - Find peers to add from selected peers (based on rating)
+            if peers_to_add_count > 0 {
+                // Take twice as many candidates as we need to add to peerlist to be able to
+                // choose best ones from them.
+                let mut candidates_to_search = peers_to_add_count * 2;
+
+                let mut candidates: Vec<(IpAddr, usize, f64)> =
+                    Vec::with_capacity(candidates_to_search as usize);
+
+                // Sort peers by rating
+                peer_ratings.sort_by(|a, b| b.2.partial_cmp(&a.2).unwrap());
+
+                // Add peers with highest rating as candidates
+                for peer in peer_ratings.iter() {
+                    // Check if peer is already in peerlist - if so go to next one
+                    if peer_list_entry.list.contains(&peer.0) {
+                        continue;
+                    }
+
+                    candidates.push(*peer);
+                    candidates_to_search -= 1;
+                    if candidates_to_search == 0 {
+                        break;
+                    }
+                }
+
+                // Here we have 2*peers_to_add_count candidates to add sorted by ranking.
+                // We need to choose best ones from them - let's choose those with lowest
+                // betweenness factor - just to avoid creating "hot" nodes that have very high
+                // importance to the network which can be risky if such node goes down.
+                candidates.sort_by(|a, b| {
+                    state.nodes[a.1]
+                        .betweenness
+                        .partial_cmp(&state.nodes[b.1].betweenness)
+                        .unwrap()
+                });
+
+                for peer in candidates.iter().take(peers_to_add_count as usize) {
+                    peer_list_entry.list.push(peer.0);
+                }
+            }
+
+            // Do not compute factors one more time after every single peerlist addition. At least
+            // for now, when computing factors is very expensive (especially betweenness and closeness).
+            // Re-calculating it after each node for whole graph would take too long.
```

Comment:
(maybe a TODO idea):
but it would be safe maybe to detect islands for adjusted peer lists to see that the algo hasn't accidentally split the network in half

Reply:
Yes, definately todo idea as currently we're using agraph to detect islands and agraph is not updated during calculations (agraph may be removed in the near future).

Reply:
Would it be okay if we leave a `TODO` comment in the code as a reminder?

Reply:
Comment left.